### PR TITLE
First attempt at guide for errors

### LINKED ISSUE #41: Convention for error messages
* Use `call. = FALSE`

* Surround names and expressions in `` ` ``, surround strings in `'`, ensure paths end in `/`

* Grammatical form

* Ideally show both what was supplied (since argument often computed and not immediately accessible) and what was expected

### PR DESCRIPTION
**Pull request by hadley**
Fixes #41

### CODE REVIEW

File: 07-errors.Rmd [hunk 1]
```diff
@@ -0,0 +1,84 @@
+# Error messages
+
+(This guide is almost entirely asprirational; most of the bad examples come from exiting tidyverse code.)
```
Comment (jennybc) [inline 07-errors.Rmd]: asprirational --> aspirational
exiting -> existing

File: 07-errors.Rmd [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+## "Must"
+
+If you can state what the user should have done to avoid the problem, start with a general statement of what was expected. Frame it possitively, and use "must". When possible, follow it up with a concise description of what was actually supplied:
```
Comment (jennybc) [inline 07-errors.Rmd]: possitively --> positively

File: 07-errors.Rmd [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+#> Error: `n` must be a numeric vector, not a character vector.
+
+dplyr::nth(1:10, 1:2)
+#> Error: `n` must have length 1, not length 2.
+
+purrr::modify_depth(list(list(x = 1)), 3, ~ . + 1)
+#> Error: `.depth` must be less than or equal to the minimum depth; 
+```
+
+Generally
```
Comment (jennybc) [inline 07-errors.Rmd]: ??

File: 07-errors.Rmd (hunk 2 above)
Comment (batpigandme) [inline 07-errors.Rmd]: I'd break that sentence up, and also add a comma after if you can, to denote conditional for stating what the user should have done vs. starting with the gen'l statement.
"If you can, state what the user should have done. Start with a general statement of what was expected."

File: 07-errors.Rmd [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+purrr::modify_depth(list(list(x = 1)), 3, ~ . + 1)
+#> Error: `.depth` must be less than or equal to the minimum depth; 
+```
+
+Generally 
+
+## Can't
+
+If you can't state what was expected, write a sentence that uses "can't"
```
Comment (batpigandme) [inline 07-errors.Rmd]: Colon?

File: 07-errors.Rmd [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+dplyr::nth(1:10, 1:2)
+#> Error: `n` must have length 1, not length 2.
+
+purrr::modify_depth(list(list(x = 1)), 3, ~ . + 1)
+#> Error: `.depth` must be less than or equal to the minimum depth; 
+```
+
+Generally 
+
+## Can't
```
Comment (krlmlr) [inline 07-errors.Rmd]: Should this be `## "Can't"` for consistency with the previous section?

File: 07-errors.Rmd [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+#> Error: Column `b` doesn't exist in `.data`
+
+pull(mtcars, b)
+#> Error: `.data` must contain column `b`.
+
+tibble(x = 1:2, y = 1:3, z = 1)
+#> Error: Column `x` must be length 1 or 3, not 2 
+```
+
+If there are multiple issues, or an inconsistency revealed across arguments or items, prefer a bulleted list:
```
Comment (krlmlr) [inline 07-errors.Rmd]: I had to double-check bulleted, it seems correct:

screenshot from 2018-01-12 22-41-07

**Comment by hadley** [comment]
Thanks for the feedback! Minor stuff fixed; and I've substantially reorganised the doc.

File: 07-errors.Rmd [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+#> `.y` has length 2
+```
+
+Prefer the use of the sigular:
+
+```{r, eval = FALSE}
+map_int(list(1, 2), ~ "a")
+
+# GOOD
+#> Error: Each result must be coercible to a single integer; result 1 is a character vector.
```
Comment (krlmlr) [inline 07-errors.Rmd]: Should this be a bullet list, too?

File: 07-errors.Rmd [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+```{r, eval = FALSE}
+mtcars %>% pull(b)
+#> Error: Can't find column `b` in `.data`
+
+as_vector(environment())
+#> Error: Can't coerece `.x` to a vector
+```
+
+## General style
```
Comment (krlmlr) [inline 07-errors.Rmd]: Maybe change caption to "Multiple error sources"?

File: 07-errors.Rmd [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+Avoid pointing the user in the wrong direction by giving an opinion about the source of the error:
```
Comment (krlmlr) [inline 07-errors.Rmd]: Is this compatible with the error we give on `filter(iris, Species = "setosa")` ?

**Comment by jennybc** [comment]
It is incredibly satisfying to see these things written down.

File: 07-errors.Rmd [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+#> * `.x` has length 4
+#> * `.y` has length 2
+
+# BAD: harder to scan
+purrr::reduce2(1:4, 1:2, `+`)
+#> Error: `.x` and `.y` must have compatible lengths: `.x` has length 4 and 
+#> `.y` has length 2
+```
+
+## Punctutation
```
Comment (krlmlr) [inline 07-errors.Rmd]: Punctuation

File: 07-errors.Rmd [hunk 11]
```diff
@@ -0,0 +1,196 @@
+# Error messages
+
+```{r}
+knitr::opts_chunk$set(eval = FALSE)
+```
```
Comment (lorenzwalthert) [inline 07-errors.Rmd]: Does this chunk need `include = FALSE` to hide the code?

File: 07-errors.Rmd [hunk 12]
```diff
@@ -86,21 +95,37 @@ purrr::reduce2(1:4, 1:2, `+`)
...
+    ```
+
+*   If you can detect multiple problems, list up to five. This allows the user
+    to fix multiple problems in a single pass without being overwhelmed by
+    many errors that may have the same source.
+
+    ```{r}
+    # BETTER
+    map_int(1:10, ~ "a")
+    #> Error: Results must be a coercible to single integers:
```
Comment (krlmlr) [inline 07-errors.Rmd]: must be ~~a~~ coercible; perhaps single -> scalar?

File: 07-errors.Rmd (hunk 11 above)
Comment (hadley) [reply 07-errors.Rmd]: Oops

**Comment by lorenzwalthert** [comment]
Ok. So then I am also going to suggest to remove the rule for `!!!` in the style guide until it's decided on how handle it.

**Comment by hadley** [comment]
I think this is good enough to merge; we can keep working on it as we notice gaps.

**Pull request merged**